internal/registry/fake.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/docker/distribution/reference"
+
+	"github.com/tetratelabs/car/internal"
+)
+
+type fakeRegistry struct {
+	ref       reference.NamedTagged
+	baseURL   string
+	fakeLayer internal.FilesystemLayer
+}
+
+// NewFake is here to keep the initial code review down
+func NewFake(ref reference.NamedTagged) internal.Registry {
+	domain := reference.Domain(ref)
+	path := reference.Path(ref)
+	baseURL := fmt.Sprintf("mem://%s/v2/%s", domain, path)
+	fakeLayer := internal.FilesystemLayer{
+		URL:       fmt.Sprintf("%s/blobs/%s", baseURL, "sha256:4e07f3bd88fb4a468d5551c21eb05f625b0efe9ee00ae25d3ffb87c0f563693f"),
+		MediaType: "application/vnd.oci.image.layer.v1.tar+gzip",
+		Size:      18500000,
+	}
+	return &fakeRegistry{ref: ref, baseURL: baseURL, fakeLayer: fakeLayer}
+}
+
+func (m *fakeRegistry) String() string {
+	return m.baseURL
+}
+
+func (m *fakeRegistry) GetImage(_ context.Context, tag, platform string) (internal.Image, error) {
+	if tag != m.ref.Tag() {
+		return internal.Image{}, fmt.Errorf("tag %s not found", tag)
+	}
+	if platform != runtime.GOOS+"/"+runtime.GOARCH {
+		return internal.Image{}, fmt.Errorf("platform %s not found", platform)
+	}
+	return internal.Image{
+		URL:              fmt.Sprintf("mem://%s/%s", tag, platform),
+		Platform:         platform,
+		FilesystemLayers: []internal.FilesystemLayer{m.fakeLayer},
+	}, nil
+}
+
+func (m *fakeRegistry) ReadFilesystemLayer(_ context.Context, layer internal.FilesystemLayer, readFile internal.ReadFile) error {
```

Comment:
I didn't backfill coverage for this as I have a working real registry and we can spend time on that. hopefully this shows how it might look, though